lib/telemetry.go

```diff
@@ -338,5 +341,78 @@ func InitTelemetry(cfg TelemetryConfig) (*metrics.InmemSink, error) {
 		metricsConf.EnableHostname = false
 		metrics.NewGlobal(metricsConf, memSink)
 	}
-	return memSink, nil
+	return sinks, errors
+}
+
+// InitTelemetry configures go-metrics based on map of telemetry config
+// values as returned by Runtimecfg.Config().
+// InitTelemetry retries configurating the sinks in case error is retriable
+// and retry_failed_connection is set to true.
+func InitTelemetry(cfg TelemetryConfig, logger hclog.Logger) (*MetricsConfig, error) {
+	if cfg.Disable {
+		return nil, nil
+	}
+
+	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
+	metrics.DefaultInmemSignal(memSink)
+	metricsConf := metrics.DefaultConfig(cfg.MetricsPrefix)
+
+	metricsConfig := &MetricsConfig{
+		Handler: memSink,
+	}
+
+	var cancel context.CancelFunc
+	var ctx context.Context
+	retryWithBackoff := func() {
+		defer func() {
+			metricsConfig.IsRetrying = false
+		}()
+
+		waiter := &retry.Waiter{
+			MaxWait: 5 * time.Minute,
+		}
+		for {
+			logger.Warn("retrying configure metric sinks", "retries", waiter.Failures())
+			_, err := configureSinks(cfg, metricsConf.HostName, memSink)
+			if err == nil {
+				logger.Info("successfully configured metrics sinks")
+				return
+			}
+			logger.Error("failed configure sinks", "error", multierror.Flatten(err))
+
+			if err := waiter.Wait(ctx); err != nil {
+				if errors.Is(err, context.Canceled) {
+					logger.Info("stop retrying configure metrics sinks")
+				} else {
+					logger.Error("waiting for retry", "error", err)
+				}
+				return
+			}
+
+		}
+	}
+
+	if _, errs := configureSinks(cfg, metricsConf.HostName, memSink); errs != nil {
+		if isRetriableError(errs) && cfg.RetryFailedConfiguration {
+			logger.Error("failed configure sinks", "error", multierror.Flatten(errs))
+			ctx, cancel = context.WithCancel(context.Background())
+			metricsConfig.IsRetrying = true
+			metricsConfig.Cancel = cancel
+			go retryWithBackoff()
+		} else {
+			return nil, errs.Unwrap()
+		}
+	}
+
+	return metricsConfig, nil
+}
+
+func isRetriableError(errs *multierror.Error) bool {
+	for _, err := range errs.WrappedErrors() {
+		var dnsError *net.DNSError
+		if errors.As(err, &dnsError) && dnsError.IsNotFound {
+			return true
+		}
+	}
```

Comment:
I think you can call `erorrs.As` with the multierror directly, rather than iterating over the wrapped errors like this.

From the [docs](https://pkg.go.dev/github.com/hashicorp/go-multierror#Error.Unwrap):

> The resulting error supports errors.As/Is/Unwrap so you can continue to use the stdlib errors package to introspect further.